Python/Product/VSInterpreters/Interpreter/WorkspaceInterpreterFactoryProvider.cs

```diff
@@ -0,0 +1,389 @@
+using Microsoft.VisualStudio.Workspace.Settings;
+using Microsoft.VisualStudio.Workspace.VSIntegration.Contracts;
+
+namespace Microsoft.PythonTools.Interpreter {
+    /// <summary>
+    /// Finds all the interpreters located under the current workspace folder,
+    /// as well as those referenced from the curent workspace folder settings.
+    /// </summary>
+    [InterpreterFactoryId(FactoryProviderName)]
+    [Export(typeof(IPythonInterpreterFactoryProvider))]
+    [Export(typeof(WorkspaceInterpreterFactoryProvider))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    class WorkspaceInterpreterFactoryProvider : IPythonInterpreterFactoryProvider, IDisposable {
+        private readonly IVsFolderWorkspaceService _workspaceService;
+        private IWorkspace _workspace;
+        private IWorkspaceSettingsManager _workspaceSettingsMgr;
+        private readonly Dictionary<string, PythonInterpreterInformation> _factories = new Dictionary<string, PythonInterpreterInformation>();
+        internal const string FactoryProviderName = WorkspaceInterpreterFactoryConstants.FactoryProviderName;
+        private FileSystemWatcher _folderWatcher;
+        private Timer _folderWatcherTimer;
+        private bool _pythonExecutableCreated;
+        private int _ignoreNotifications;
+        private bool _initialized;
+
+        private static readonly Version[] ExcludedVersions = new[] {
+            new Version(2, 5),
+            new Version(3, 0)
+        };
+
+        internal event EventHandler DiscoveryStarted;
+
+        [ImportingConstructor]
+        public WorkspaceInterpreterFactoryProvider(
+            [Import] IVsFolderWorkspaceService workspaceService = null
+        ) {
+            _workspaceService = workspaceService;
+            _workspaceService.OnActiveWorkspaceChanged += OnActiveWorkspaceChanged;
+        }
+
+        protected void Dispose(bool disposing) {
+            if (disposing) {
+                _workspaceService.OnActiveWorkspaceChanged -= OnActiveWorkspaceChanged;
+                if (_workspaceSettingsMgr != null) {
+                    _workspaceSettingsMgr.OnWorkspaceSettingsChanged -= OnSettingsChanged;
+                }
+                _folderWatcher?.Dispose();
+                _folderWatcherTimer?.Dispose();
+            }
+        }
+
+        public void Dispose() {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        ~WorkspaceInterpreterFactoryProvider() {
+            Dispose(false);
+        }
+
+        private void EnsureInitialized() {
+            lock (_factories) {
+                if (!_initialized) {
+                    _initialized = true;
+                    InitializeWorkspace();
+                    DiscoverInterpreterFactories();
+                }
+            }
+        }
+
+        private Task OnActiveWorkspaceChanged(object sender, EventArgs e) {
+            InitializeWorkspace();
+            DiscoverInterpreterFactories();
```

Comment:
`InitializeWorkspace` and `DiscoverInterpreterFactories` are called without lock here but under lock in `EnsureInitialized`. Is that expected?

Reply:
You're right, there are some missing locks